File: src/main/java/org/flossware/civilization/model/ClimateState.java
package org.flossware.civilization.model;

/**
 * Multi-dimensional climate state.
 */
public record ClimateState(
    double temperatureAnomaly,
    double droughtIndex,
    double stormFrequency,
    double seaLevelRise_mm
) {
    public ClimateState {
        if (temperatureAnomaly < -10 || temperatureAnomaly > 10) {
            throw new IllegalArgumentException("Temperature anomaly out of range [-10, 10]: " + temperatureAnomaly);
        }
        if (droughtIndex < 0 || droughtIndex > 1) {
            throw new IllegalArgumentException("Drought index must be in [0, 1]");
        }
        if (stormFrequency < 0) {
            throw new IllegalArgumentException("Storm frequency cannot be negative");
        }
    }

    public ClimateState withTemperatureAnomaly(double newAnomaly) {
        return new ClimateState(newAnomaly, droughtIndex, stormFrequency, seaLevelRise_mm);
    }

    public ClimateState withDroughtIndex(double newIndex) {
        return new ClimateState(temperatureAnomaly, newIndex, stormFrequency, seaLevelRise_mm);
    }

    public ClimateState withStormFrequency(double newFrequency) {
        return new ClimateState(temperatureAnomaly, droughtIndex, newFrequency, seaLevelRise_mm);
    }

    public ClimateState withSeaLevelRise(double newRise) {
        return new ClimateState(temperatureAnomaly, droughtIndex, stormFrequency, newRise);
    }

    public double getResourceAbundance() {
        double droughtPenalty = Math.abs(droughtIndex - 0.5) * 2;
        double tempPenalty = Math.abs(temperatureAnomaly) * 0.1;
        return Math.max(0.1, 1.0 - droughtPenalty * 0.3 - tempPenalty);
    }
}

File: mobile/shared/src/commonMain/kotlin/org/flossware/civilization/model/ClimateState.kt
package org.flossware.civilization.model

import kotlinx.serialization.Serializable
import kotlin.math.abs
import kotlin.math.max

@Serializable
data class ClimateState(
    val temperatureAnomaly: Double,
    val droughtIndex: Double,
    val stormFrequency: Double,
    val seaLevelRise_mm: Double
) {
    init {
        require(temperatureAnomaly in -10.0..10.0) { "Temperature anomaly out of range [-10, 10]: $temperatureAnomaly" }
        require(droughtIndex in 0.0..1.0) { "Drought index must be in [0, 1]" }
        require(stormFrequency >= 0) { "Storm frequency cannot be negative" }
    }

    fun withTemperatureAnomaly(newAnomaly: Double) = copy(temperatureAnomaly = newAnomaly)
    fun withDroughtIndex(newIndex: Double) = copy(droughtIndex = newIndex)
    fun withStormFrequency(newFrequency: Double) = copy(stormFrequency = newFrequency)
    fun withSeaLevelRise(newRise: Double) = copy(seaLevelRise_mm = newRise)

    fun getResourceAbundance(): Double {
        val droughtPenalty = abs(droughtIndex - 0.5) * 2
        val tempPenalty = abs(temperatureAnomaly) * 0.1
        return max(0.1, 1.0 - droughtPenalty * 0.3 - tempPenalty)
    }
}
